File: analysis/02_fpi_calibration/fpi_calibration.R
# Is ESPN FPI worth anything against the DraftKings closer?
#
#   A. Calibration — binned FPI home win prob vs observed home win rate (Wilson CI).
#   B. Accuracy — RMSE of actual margin vs FPI margin, vs the closer, vs a 50/50 blend.
#   C. ATS — when FPI and the closer disagree by |delta| pts, how often does the
#      FPI side cover? Compare to the 52.4% break-even at -110.
#
# Output: analysis/_out/fpi_calibration.csv, fpi_ats_by_delta.csv.
# Mirrors fpi_calibration.py — keep them in lockstep.

suppressPackageStartupMessages(library(dplyr))

.this_dir <- tryCatch(dirname(sys.frame(1)$ofile), error = function(e) "analysis/02_fpi_calibration")
source(file.path(.this_dir, "..", "_shared", "load_data.R"))

OUT_DIR      <- normalizePath(file.path(.this_dir, "..", "_out"), mustWork = FALSE)
PROB_BINS    <- c(0, 0.2, 0.4, 0.6, 0.8, 1.0001)
DELTA_BINS   <- c(0, 3, 5, 8, 99)
DELTA_LABELS <- c("0-3", "3-5", "5-8", "8+")
dir.create(OUT_DIR, showWarnings = FALSE, recursive = TRUE)

rmse <- function(a, b) sqrt(mean((a - b)^2))

g <- load_games()
if (nrow(g) == 0) {
  cat("no settled games yet — run cfb_edge.py --backfill --date <past saturday>\n")
  quit(status = 0)
}
cat(sprintf("Settled FBS-vs-FBS games with closer + FPI: %s\n\n", format(nrow(g), big.mark = ",")))

# A. calibration — same left-closed bins as pandas cut(right=False)
cat("A. FPI home win-prob calibration (Wilson 95% CI)\n")
g$pbin <- cut(g$home_fpi_p, PROB_BINS, right = FALSE)
cal <- g %>% group_by(pbin) %>% group_map(function(grp, key) {
  w <- wilson(sum(grp$home_won), nrow(grp))
  lab <- sub("\\)$", ")", sub("^\\[", "[", as.character(key$pbin)))
  cat(sprintf("  %-12s n=%4d  pred %5.1f%%  obs %5.1f%% [%5.1f,%5.1f]  Δ%+5.1fpp\n",
              lab, nrow(grp), 100 * mean(grp$home_fpi_p), 100 * w[1], 100 * w[2], 100 * w[3],
              100 * (w[1] - mean(grp$home_fpi_p))))
  data.frame(bin = lab, n = nrow(grp), pred_mean = mean(grp$home_fpi_p),
             obs_rate = w[1], obs_lo = w[2], obs_hi = w[3])
}) %>% bind_rows()
write.csv(cal, file.path(OUT_DIR, "fpi_calibration.csv"), row.names = FALSE)

# B. accuracy
blend   <- 0.5 * g$home_fpi_margin + 0.5 * g$market_margin
r_fpi   <- rmse(g$home_margin, g$home_fpi_margin)
r_mkt   <- rmse(g$home_margin, g$market_margin)
r_blend <- rmse(g$home_margin, blend)
cat("\nB. Margin accuracy (RMSE, lower is better)\n")
cat(sprintf("  FPI margin   %6.2f\n  DK closer    %6.2f\n  50/50 blend  %6.2f\n", r_fpi, r_mkt, r_blend))
cat(sprintf("  verdict: %s by %.2f pts RMSE (MARGIN_SD in cfb_edge.py assumes ~13.5)\n",
            if (r_fpi < r_mkt) "FPI beats the closer" else "the closer beats FPI", abs(r_fpi - r_mkt)))

# C. ATS by |delta|
cat(sprintf("\nC. FPI-side cover rate by |FPI − closer| (break-even at -110 = %.1f%%)\n", 100 * BREAK_EVEN_110))
d <- g[!is.na(g$fpi_side_covered), ]
d$dbin <- cut(d$abs_delta, DELTA_BINS, labels = DELTA_LABELS, right = FALSE)
ats <- d %>% group_by(dbin) %>% group_map(function(grp, key) {
  w <- wilson(sum(grp$fpi_side_covered), nrow(grp))
  v <- if (w[2] > BREAK_EVEN_110) "PROFITABLE (CI lo > BE)" else if (w[3] < BREAK_EVEN_110) "losing (CI hi < BE)" else "inconclusive"
  cat(sprintf("  Δ%-5s n=%4d  cover %5.1f%% [%5.1f,%5.1f]  %s\n",
              as.character(key$dbin), nrow(grp), 100 * w[1], 100 * w[2], 100 * w[3], v))
  data.frame(delta_bin = as.character(key$dbin), n = nrow(grp), covers = sum(grp$fpi_side_covered),
             cover_rate = w[1], ci_lo = w[2], ci_hi = w[3], verdict = v)
}) %>% bind_rows()
w <- wilson(sum(d$fpi_side_covered), nrow(d))
cat(sprintf("  ALL    n=%4d  cover %5.1f%% [%5.1f,%5.1f]\n", nrow(d), 100 * w[1], 100 * w[2], 100 * w[3]))
write.csv(ats, file.path(OUT_DIR, "fpi_ats_by_delta.csv"), row.names = FALSE)
cat(sprintf("\nwrote %s, %s\n", file.path(OUT_DIR, "fpi_calibration.csv"), file.path(OUT_DIR, "fpi_ats_by_delta.csv")))

File: analysis/02_fpi_calibration/fpi_calibration.py
"""Is ESPN FPI worth anything against the DraftKings closer?

Three questions, all on EVERY settled FBS-vs-FBS game (not just flagged ones):
  A. Calibration — binned FPI home win prob vs observed home win rate (Wilson CI).
  B. Accuracy — RMSE of actual margin vs FPI margin, vs the closing spread, vs a
     50/50 blend. If the closer's RMSE is lower, the market is the better model.
  C. ATS — when FPI and the closer disagree by |delta| pts, how often does the
     FPI side cover? Compare to the 52.4% break-even at -110.

Output: analysis/_out/fpi_calibration.csv, fpi_ats_by_delta.csv.
Mirrors fpi_calibration.R — keep them in lockstep.
"""
from __future__ import annotations

import sys
from pathlib import Path

import numpy as np
import pandas as pd

sys.path.insert(0, str(Path(__file__).resolve().parents[1]))
from _shared.load_data import BREAK_EVEN_110, load_games, wilson  # noqa: E402

OUT_DIR = Path(__file__).resolve().parents[1] / "_out"
PROB_BINS = [0.0, 0.2, 0.4, 0.6, 0.8, 1.0001]
DELTA_BINS = [0.0, 3.0, 5.0, 8.0, 99.0]
DELTA_LABELS = ["0-3", "3-5", "5-8", "8+"]


def rmse(a: pd.Series, b: pd.Series) -> float:
    return float(np.sqrt(((a - b) ** 2).mean()))


def main() -> None:
    OUT_DIR.mkdir(exist_ok=True)
    g = load_games()
    if g.empty:
        print("no settled games yet — run cfb_edge.py --backfill --date <past saturday>")
        return
    print(f"Settled FBS-vs-FBS games with closer + FPI: {len(g):,}\n")

    # A. calibration
    print("A. FPI home win-prob calibration (Wilson 95% CI)")
    g["pbin"] = pd.cut(g.home_fpi_p, PROB_BINS, right=False)
    rows = []
    for b, grp in g.groupby("pbin", observed=True):
        p, lo, hi = wilson(int(grp.home_won.sum()), len(grp))
        rows.append({"bin": str(b), "n": len(grp), "pred_mean": grp.home_fpi_p.mean(),
                     "obs_rate": p, "obs_lo": lo, "obs_hi": hi})
        print(f"  {str(b):<12} n={len(grp):>4}  pred {100*grp.home_fpi_p.mean():5.1f}%  "
              f"obs {100*p:5.1f}% [{100*lo:5.1f},{100*hi:5.1f}]  Δ{100*(p-grp.home_fpi_p.mean()):+5.1f}pp")
    pd.DataFrame(rows).to_csv(OUT_DIR / "fpi_calibration.csv", index=False)

    # B. accuracy
    blend = 0.5 * g.home_fpi_margin + 0.5 * g.market_margin
    r_fpi, r_mkt, r_blend = rmse(g.home_margin, g.home_fpi_margin), rmse(g.home_margin, g.market_margin), rmse(g.home_margin, blend)
    print("\nB. Margin accuracy (RMSE, lower is better)")
    print(f"  FPI margin   {r_fpi:6.2f}\n  DK closer    {r_mkt:6.2f}\n  50/50 blend  {r_blend:6.2f}")
    print(f"  verdict: {'FPI beats the closer' if r_fpi < r_mkt else 'the closer beats FPI'} "
          f"by {abs(r_fpi - r_mkt):.2f} pts RMSE (MARGIN_SD in cfb_edge.py assumes ~13.5)")

    # C. ATS by |delta|
    print(f"\nC. FPI-side cover rate by |FPI − closer| (break-even at -110 = {100*BREAK_EVEN_110:.1f}%)")
    d = g[g.fpi_side_covered.notna()].copy()
    d["dbin"] = pd.cut(d.abs_delta, DELTA_BINS, labels=DELTA_LABELS, right=False)
    rows = []
    for b, grp in d.groupby("dbin", observed=True):
        p, lo, hi = wilson(int(grp.fpi_side_covered.sum()), len(grp))
        verdict = "PROFITABLE (CI lo > BE)" if lo > BREAK_EVEN_110 else "losing (CI hi < BE)" if hi < BREAK_EVEN_110 else "inconclusive"
        rows.append({"delta_bin": b, "n": len(grp), "covers": int(grp.fpi_side_covered.sum()),
                     "cover_rate": p, "ci_lo": lo, "ci_hi": hi, "verdict": verdict})
        print(f"  Δ{b:<5} n={len(grp):>4}  cover {100*p:5.1f}% [{100*lo:5.1f},{100*hi:5.1f}]  {verdict}")
    p, lo, hi = wilson(int(d.fpi_side_covered.sum()), len(d))
    print(f"  ALL    n={len(d):>4}  cover {100*p:5.1f}% [{100*lo:5.1f},{100*hi:5.1f}]")
    pd.DataFrame(rows).to_csv(OUT_DIR / "fpi_ats_by_delta.csv", index=False)
    print(f"\nwrote {OUT_DIR / 'fpi_calibration.csv'}, {OUT_DIR / 'fpi_ats_by_delta.csv'}")


if __name__ == "__main__":
    main()
